capsules/extra/src/net/thread/thread_utils.rs

```diff
@@ -0,0 +1,282 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+use crate::net::stream::{encode_bytes, SResult};
+use crate::net::thread::tlv::{unwrap_tlv_offset, LinkMode, MulticastResponder, Tlv};
+use crate::net::{ieee802154::MacAddress, ipv6::ip_utils::IPAddr};
+
+use kernel::ErrorCode;
+
+pub const SECURITY_SUITE_LEN: usize = 1;
+pub const AUX_SEC_HEADER_LENGTH: usize = 10;
+pub const AUTH_DATA_LEN: usize = 42;
+pub const IPV6_LEN: usize = 16;
+const PARENT_REQUEST_MLE_SIZE: usize = 21;
+pub const MULTICAST_IPV6: IPAddr = IPAddr([
+    0xff, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02,
+]);
+
+#[derive(Clone, Copy)]
+pub struct NetworkKey {
+    pub mle_key: [u8; 16],
+    pub mac_key: [u8; 16],
+}
+
+pub enum ThreadState {
+    SendParentReq,
+    WaitingParentRsp,
+    RecvParentRsp(IPAddr),
+    SendChildIdReq(IPAddr, MacAddress),
+    WaitingChildRsp,
+    RecvChildRsp(IPAddr),
+    SEDActive(IPAddr, MacAddress),
+    SendUpdate(IPAddr, MacAddress),
+    SendUDPMsg,
+    Detached,
+}
+
+pub enum MleCommand {
+    LinkRequest = 0,
+    LinkAccept = 1,
+    LinkAcceptAndRequest = 2,
+    LinkAdvertisement = 4,
+    DataRequest = 7,
+    DataResponse = 8,
+    ParentRequest = 9,
+    ParentResponse = 10,
+    ChildIdRequest = 11,
+    ChildIdResponse = 12,
+    ChildUpdateRequest = 13,
+    ChildUpdateResponse = 14,
+    Announce = 15,
+    DiscoverRequest = 16,
+    DiscoveryResponse = 17,
+    LinkMetricsManagReq = 18,
+    LinkMetricsManagResp = 19,
+    LinkProbe = 20,
+}
+
+pub fn generate_src_ipv6(macaddr: &[u8; 8]) -> IPAddr {
+    // -----------------------------------------------------------------------------------------------
+    // THREAD SPEC 5.2.2.4 (V1.3.0) -- A Thread Device MUST assign a link-local IPv6 address where the
+    // interface identifier is set to the MAC Extended Address with the universal/local bit inverted.
+    // ------------------------------------------------------------------------------------------------
+
+    let mut output: [u8; 16] = [0; 16];
+    let mut lower_bytes = *macaddr;
+
+    // The universal/local bit is the 2nd least significant bit.
+    // Invert by xor first byte of MAC addr with 2
+    lower_bytes[0] = macaddr[0] ^ 2;
+    let upper_bytes: [u8; 8] = [0xfe, 0x80, 0, 0, 0, 0, 0, 0];
+
+    encode_bytes(&mut output[..8], &upper_bytes);
+    encode_bytes(&mut output[8..16], &lower_bytes);
+    IPAddr(output)
+}
+
+pub fn mac_from_ipv6(ipv6: IPAddr) -> [u8; 8] {
+    // Helper function to generate the mac address from the mac address;
+    // reversing the tranformation used/described in `generate_src_ipv6`
```

Comment:
```suggestion
/// Helper function to generate the mac address from the mac address;
/// the tranformation used/described in `generate_src_ipv6`
pub fn mac_from_ipv6(ipv6: IPAddr) -> [u8; 8] {
```

Here and elsewhere, promote these to doc comments [especially important for `pub` functions]